### PR TITLE
ARROW-4957: [Rust] [DataFusion] Re-implement get_supertype 

### PR DESCRIPTION
**Pull request by andygrove**
This PR re-implements get_supertype to be easier to read and maintain. It still only supports numeric types and string. After this PR is merged I will create new PRs to add rules for boolean and date/time.

### CODE REVIEW

**Comment by andygrove** [comment]
@kszucs Please take a look when you can

**Comment by andygrove** [comment]
@paddyhoran @sunchao Some code cleanup that would be good to get into 0.13

**Review by uagashe-rms — APPROVED** [review]
lgtm

File: rust/datafusion/src/optimizer/utils.rs [hunk 1]
```diff
@@ -98,117 +98,117 @@ pub fn exprlist_to_fields(expr: &Vec<Expr>, input_schema: &Schema) -> Result<Vec
...
-        Some(dt) => Ok(dt),
-        None => match _get_supertype(r, l) {
-            Some(dt) => Ok(dt),
-            None => Err(ExecutionError::InternalError(format!(
-                "Failed to determine supertype of {:?} and {:?}",
-                l, r
-            ))),
-        },
+    use arrow::datatypes::DataType::*;
+    let d = if l == r {
```
Comment (sunchao) [inline rust/datafusion/src/optimizer/utils.rs]: In the old implementation the super type of `UInt16` and `Int8` is `None` but now it is `Int16`. Why is that? Also do you know how C++ handles this case?

File: rust/datafusion/src/optimizer/utils.rs [hunk 2]
```diff
@@ -217,9 +217,133 @@ mod tests {
...
+                result.push_str(&format!(
+                    "supertype of {:?} and {:?} is {:?}
+\n",
+                    l,
+                    r,
+                    get_supertype(l, r).unwrap()
+                ));
+            }
+        }
+        assert_eq!(
```
Comment (sunchao) [inline rust/datafusion/src/optimizer/utils.rs]: Can we avoid comparing with string? whenever there's an error it is very hard to find out since you are comparing two huge strings.

**Review by sunchao — COMMENTED** [review]
Thanks @andygrove . Sorry for the late review.

File: rust/datafusion/src/optimizer/utils.rs (hunk 1 above)
Comment (fsaintjacques) [reply rust/datafusion/src/optimizer/utils.rs]: It is not statically compatible in C++ (safely). https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/cast.cc#L156-L208

It can be safe to cast, but it needs a runtime check.

File: rust/datafusion/src/optimizer/utils.rs [hunk 3]
```diff
@@ -98,117 +98,117 @@ pub fn exprlist_to_fields(expr: &Vec<Expr>, input_schema: &Schema) -> Result<Vec
...
-                "Failed to determine supertype of {:?} and {:?}",
-                l, r
-            ))),
-        },
+    use arrow::datatypes::DataType::*;
+    let d = if l == r {
+        Some(l.clone())
+    } else if l == &Utf8 || r == &Utf8 {
+        Some(Utf8)
+    } else if is_signed_int(l) {
```
Comment (fsaintjacques) [inline rust/datafusion/src/optimizer/utils.rs]: This block is hard to follow, I'd rewrite this with small functions.

**Comment by wesm** [comment]
Can this PR be merged?

**Comment by sunchao** [comment]
I think there are a few pending comments to be resolved. @andygrove can you take a look?

**Comment by andygrove** [comment]
Sorry, I haven't had any spare time for a while due to work commitments. I am closing this and will try again another time.
